README.md
# aiosenseme library

[![PyPI version](https://badge.fury.io/py/aiosenseme.svg)](https://badge.fury.io/py/aiosenseme) [![Downloads](https://pepy.tech/badge/aiosenseme)](https://pepy.tech/project/aiosenseme)

This branch of aiosenseme is intended to help test i6 fans which appears to not work like other SenseME fans.

## Command line access

To access a fan directly by ip address skipping the discovery step. Name must match the one you gave the fan in the Haiku app.

```console
$ aiosenseme --debug --name "Studio Beam Fan" --ip 172.20.3.39
DEBUG:aiosenseme.fan:Studio Beam Fan: Connecting
DEBUG:aiosenseme.fan:Studio Beam Fan: Connected
DEBUG:aiosenseme.fan:Studio Beam Fan: Status update
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [SNSROCC;STATUS]='OCCUPIED'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [DEVICE;BEEPER]='ON'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [DEVICE;INDICATORS]='ON'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [DEVICE;LIGHT]='PRESENT'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [DEVICE;SERVER]='PRODUCTION'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [ERRORLOG;ENTRIES;NUM]='10'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [ERRORLOG;ENTRIES;MAX]='10'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [FAN;AUTO]='OFF'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [FAN;DIR]='FWD'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [FAN;PWR]='ON'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [FAN;SPD;ACTUAL]='1'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [FAN;SPD;MAX]='7'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [FAN;SPD;MIN]='1'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [FAN;BOOKENDS]='1;7'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [FAN;TIMER;CURR]='0'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [FAN;TIMER;MAX]='7'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [FAN;TIMER;MIN]='1'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [FAN;WHOOSH;STATUS]='OFF'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [FW;NAME]='FW000003'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [FW;FW000003]='2.5.0'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [GROUP;LIST]='Studio Fan'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [GROUP;ROOM;TYPE]='6'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [LEARN;MAXSPEED]='7'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [LEARN;MINSPEED]='0'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [LEARN;ZEROTEMP]='2388'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [LIGHT;AUTO]='OFF'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [LIGHT;LEVEL;ACTUAL]='0'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [LIGHT;BOOKENDS]='1;16'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [LIGHT;LEVEL;MAX]='16'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [LIGHT;LEVEL;MIN]='1'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [LIGHT;PWR]='OFF'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [NAME;VALUE]='Studio Beam Fan'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [NW;AP;STATUS]='OFF'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [NW;DHCP]='OFF'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [NW;PARAMS;ACTUAL]='172.20.3.39;255.255.0.0;172.20.0.1'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [NW;SSID]='Lawrence b'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [NW;TOKEN]='<redacted>'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [SCHEDULE;CAP]='MAX EVENTS,29'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [SCHEDULE;EVENT;LIST]='NONE'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [SLEEP;EVENT;OFF]='LIGHT,PWR,OFF'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [SLEEP;EVENT]='ON'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [SLEEP;STATE]='OFF'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [SMARTMODE;ACTUAL]='OFF'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [SMARTMODE;STATE]='OFF'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [SMARTSLEEP;IDEALTEMP]='2111'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [SMARTSLEEP;MAXSPEED]='7'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [SMARTSLEEP;MINSPEED]='0'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [SNSROCC;TIMEOUT;CURR]='660000'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [SNSROCC;TIMEOUT;MAX]='86400000'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [SNSROCC;TIMEOUT;MIN]='60000'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [WINTERMODE;HEIGHT]='213'
DEBUG:aiosenseme.fan:Studio Beam Fan: Param updated: [WINTERMODE;STATE]='OFF'
```

This is what is looks like if an incorrect name is specified.

```console
$ aiosenseme --debug --name "Unknown Name" --ip 172.20.3.39
DEBUG:aiosenseme.fan:Unknown Name: Connecting
DEBUG:aiosenseme.fan:Unknown Name: Connected
DEBUG:aiosenseme.fan:Unknown Name: Status update
Unknown Name
  Model: Haiku Fan without light, FW Version: None
  IP Addr: 172.20.3.39, MAC Addr: 01:23:45:67:89:AB
  Token: None
State: Fan is off, Light is off, Whoosh: off
DEBUG:aiosenseme.fan:Unknown Name: Listener task cancelled
DEBUG:aiosenseme.fan:Unknown Name: Updater task cancelled
```

## Issues

* Do use this for anything other than testing i6 fans.


CHANGELOG.md
# Changelog for aiosenseme library

## v0.3.4-i6 - Test i6 fans only

* Please do not use this for normal SenseME fans.

## v0.3.3

* Ignore Wireless Wall Controllers. It appears this library will connect to them and cause lockups requiring a reset.
* Add error logging for unhandled exceptions in SensemeDiscovery._updater() and SensemeFan._listener().

## v0.3.2

* Add H-Series Haiku fan to known model list. Fan reports a slightly different model string for the Haiku Fan.
* Add model list to command line arguments.
* Add exception handling when creating discovery sockets. This should help the issue of a socket error on one network interface preventing discovery from working other network interfaces.

## v0.3.1

* SensemeFan.room_type now reports the room correctly.
* Removed socket option reuse_port for discovery as is appears to not work on MacOS and really isn't needed.
* Discovery no longer listens on loopback interface.

## v0.3.0

* Lots of linting changes.
* API breaking changes.
  * aiosenseme.Discover() is now aiosenseme.discover()
  * aiosenseme.Discover_Any() is now aiosenseme.discover_any()
* aiosenseme.discover() will now match Fan Name, Room Name and IP address

## v0.2.1

* Requiring Python 3.7 was too restrictive. Back to allowing and testing Python 3.6.

## v0.2.0

* API breaking changes.
  * SensemeFan.group_status() is now SensemeFan.room_status()
  * SensemeFan.group_name() is now SensemeFan.room_name()
  * SensemeFan.group_room_type() is now SensemeFan.room_type()
* Now requiring python 3.7 or above. Not testing on anything lower.
* Add Discover() function to aiosenseme.
* Improved error handling in SensemeDiscovery and some testing on MacOS.
* Handle fan models differently and hopefully better in SensemeFan.
* Added SensemeFan.models() class method which returns an list of known models.
* API Added command line script capability.
* Added Type Hints.

## v0.1.1

* Add L-Series Haiku fan to known model list.

## v0.1.0

* Initial release.


aiosenseme/discovery.py
"""SenseME Discovery.

This class periodically broadcasts discovery packets and listens for response messages
from SenseME fan by Big Ass Fans. Discovered fans are announced via callbacks.

Based on work from Bruce at http://bruce.pennypacker.org/tag/senseme-plugin/
and https://github.com/bpennypacker/SenseME-Indigo-Plugin

Based on work from TomFaulkner at https://github.com/TomFaulkner/SenseMe

Source can be found at https://github.com/mikelawrence/aiosenseme
"""
import asyncio
import inspect
import ipaddress
import logging
import random
import socket
import time
import traceback

import ifaddr

from .fan import SensemeFan

_LOGGER = logging.getLogger(__name__)

PORT = 31415


class SensemeDiscoveryEndpoint:
    """High-level endpoint for SenseME Discovery protocol."""

    receive_queue = asyncio.Queue()

    def __init__(self, ip: str = None):
        """Initialize Senseme Discovery Endpoint."""
        self.opened = False
        self.transport = None
        self.ip = ip

    def abort(self):
        """Close the transport immediately.

        Buffered write data will be lost.
        """
        self.opened = False
        if self.transport is None:
            return
        self.transport.abort()
        self.close()

    def close(self):
        """Close the transport gracefully.

        Buffered write data will be sent.
        """
        self.opened = False
        if self.transport is None:
            return
        self.receive_queue.put_nowait(None)  # tell receive() socket is closed
        if self.transport:
            self.transport.close()

    def is_closing(self) -> bool:
        """Return True if the endpoint is closed or closing."""
        if not self.opened:
            return True  # unopened connection is closed
        if self.transport is None:
            return True  # opened connection but no transport is closed
        return self.transport.is_closing()

    async def receive(self) -> SensemeFan:
        """Wait for discovered device and return it.

        Return None when the socket is closed.
        This method is a coroutine.
        """
        while True:
            if self.receive_queue.empty() and self.transport.is_closing():
                return None
            rsp = await self.receive_queue.get()
            if rsp is None:
                return None
            msg = rsp[0]
            addr = rsp[1]
            _LOGGER.debug("Received '%s' from %s on %s", msg, addr, self.ip)
            if len(msg) > 200 or len(msg) < 31:
                continue
            if msg[0] != "(" or msg[-1] != ")":
                continue
            msg = msg[1:-1]
            msg_data = msg.split(";")
            if len(msg_data) != 5:
                continue
            if msg_data[4] == "SWITCH,SENSEME":
                _LOGGER.debug("Ignored Wall Switch", msg[0], addr)
                continue
            device = SensemeFan(msg_data[0], msg_data[3], addr, msg_data[4])
            return device

    def send_broadcast(self):
        """Send the SenseME Discovery broadcast packet."""
        if not self.is_closing():
            data = "<ALL;DEVICE;ID;GET>".encode("utf-8")
            self.transport.sendto(data, ("<broadcast>", PORT))
            _LOGGER.debug("Discovery broadcast on %s", self.ip)


class SensemeDiscoveryProtocol(asyncio.DatagramProtocol):
    """Datagram protocol for SenseME Discovery."""

    def __init__(self, endpoint: SensemeDiscoveryEndpoint):
        """Initialize Senseme Discovery Protocol."""
        self._endpoint = endpoint

    # Protocol methods
    def connection_made(self, transport: asyncio.Protocol):
        """Socket connect on SenseME Discovery Protocol."""
        self._endpoint.transport = transport
        self._endpoint.opened = True
        _LOGGER.debug("Listening on %s", self._endpoint.ip)
        self._endpoint.send_broadcast()

    def connection_lost(self, exc):  # pylint: disable=unused-argument
        """Lost connection SenseME Discovery Protocol."""
        _LOGGER.debug("Listener closed on %s", self._endpoint.ip)
        self._endpoint.close()  # half-closed connections are not permitted

    # Datagram protocol methods
    def datagram_received(self, data: str, addr: str):
        """UDP packet received on SenseME Discovery Protocol."""
        if data:
            msg = data.decode("utf-8")
            try:
                self._endpoint.receive_queue.put_nowait((msg, addr[0]))
            except asyncio.QueueFull:
                _LOGGER.error("Receive queue full")

    def error_received(self, exc):
        """Error on SenseME Discovery Protocol."""
        _LOGGER.debug(
            "Protocol error on %s\n%s", self._endpoint.ip, traceback.format_exc()
        )
        self._endpoint.close()


class SensemeDiscovery:
    """SenseME Discovery Class.

    This class periodically broadcasts discovery packets and listens
    for response messages from SenseME fans by Big Ass Fans.
    """

    _devices = []  # all SensemeDiscovery objects use the same device list

    def __init__(self, start_first: bool = True, refresh_minutes: int = 5):
        """Initialize Senseme Discovery Protocol."""
        self.start_first = start_first
        self.refresh_minutes = refresh_minutes
        self._is_running = False
        self._callbacks = []
        self._broadcaster_task = None

    @property
    def devices(self):
        """Get the current list of discovered fans."""
        return self._devices

    def add_callback(self, callback):
        """Add callback function/coroutine.

        Called when parameters are updated.
        """
        if callback not in self._callbacks:
            self._callbacks.append(callback)
            if inspect.iscoroutinefunction(callback):
                asyncio.get_event_loop().create_task(callback(self._devices.copy()))
                _LOGGER.debug("Added coroutine callback")
            else:
                callback(self._devices.copy())
                _LOGGER.debug("Added function callback")

    def remove_callback(self, callback):
        """Remove existing callback function/coroutine."""
        if callback in self._callbacks:
            self._callbacks.remove(callback)
            _LOGGER.debug("Removed callback")

    async def _create_endpoints(self):
        """Create an endpoint per interface and return as a list of endpoints."""
        endpoints = []
        loop = asyncio.get_event_loop()
        listening = 0
        for adapter in ifaddr.get_adapters():
            for ip in adapter.ips:
                if ip.is_IPv4 and not ipaddress.ip_address(ip.ip).is_loopback:
                    endpoint = SensemeDiscoveryEndpoint(ip.ip)
                    # _LOGGER.debug("Found IPv4 %s", ip.ip)
                    try:
                        await loop.create_datagram_endpoint(
                            lambda ep=endpoint: SensemeDiscoveryProtocol(ep),
                            local_addr=(ip.ip, PORT),
                            family=socket.AF_INET,
                            allow_broadcast=True,
                        )
                        endpoints.append(endpoint)
                        listening += 1
                    except OSError:
                        last_error = (
                            f"Create datagram endpoint error on {ip.ip}\n"
                            f"{traceback.format_exc()}"
                        )
                        _LOGGER.debug(last_error)
        if listening == 0:
            # failed to bind to any address
            _LOGGER.error(
                "Failed to listen for discovery responses on any address. "
                "Last error\n%s",
                last_error,
            )

        return endpoints

    async def _broadcaster(self):
        """Periodically broadcast discovery packet.

        If the underlying socket has an error this task will exit.
        This method is a coroutine.
        """
        self._is_running = True
        loop = asyncio.get_event_loop()
        device = None
        while True:
            try:
                found_new = 0
                found_total = 0
                found_devices = []
                endpoints = await self._create_endpoints()
                start = time.time()
                while True:
                    try:
                        device = await asyncio.wait_for(endpoints[0].receive(), 1)
                    except asyncio.TimeoutError:
                        device = None
                        if time.time() - start < 5:
                            for endpoint in endpoints:
                                endpoint.send_broadcast()
                        else:
                            for endpoint in endpoints:
                                endpoint.abort()
                            endpoints = []
                            break
                    if device is not None:
                        if device not in self._devices:
                            if self.start_first:
                                if await device.update():
                                    self._devices.append(device)
                                    _LOGGER.debug("Discovered %s", device)
                                    found_new += 1
                                else:
                                    _LOGGER.debug("Failed to start %s", device.name)
                            else:
                                if await device.fill_out_secondary_info():
                                    self._devices.append(device)
                                    _LOGGER.debug("Discovered %s", device)
                                    found_new += 1
                                else:
                                    _LOGGER.debug(
                                        "Failed to retrieve secondary info for %s",
                                        device.name,
                                    )
                        if device not in found_devices:
                            found_devices.append(device)
                            found_total += 1
                        for callback in self._callbacks:
                            if inspect.iscoroutinefunction(callback):
                                loop.create_task(callback(self._devices.copy()))
                            else:
                                callback(self._devices.copy())
                found_old = found_total - found_new
                found_devices = None
                if found_old > 1:
                    _LOGGER.debug(
                        "Discovered %s existing fan%s",
                        found_old,
                        "" if found_old == 1 else "s",
                    )
                elif found_new == 0:
                    _LOGGER.debug("Discovered 0 fans")
                await asyncio.sleep(self.refresh_minutes * 60 + random.uniform(-10, 10))
            except asyncio.CancelledError:
                _LOGGER.debug("Broadcaster task cancelled")
                return
            except Exception:
                _LOGGER.error("Broadcaster task error\n%s", traceback.format_exc())
                raise
            finally:
                for endpoint in endpoints:
                    endpoint.abort()
                endpoints = None
                found_devices = None
        _LOGGER.error("Broadcaster task ended")

    def start(self):
        """Start both broadcaster and listener tasks.

        Will maintain a list of discovered fans.
        """
        if not self._is_running:
            loop = asyncio.get_event_loop()
            self._broadcaster_task = loop.create_task(self._broadcaster())

    def stop(self):
        """Stop both broadcaster and listener tasks.

        Any discovered fans will remain in memory and will continue to update.
        """
        if self._is_running is True:
            self._broadcaster_task.cancel()
            self._is_running = False

    def remove_discovered_devices(self):
        """Stop both broadcaster and listener tasks.

        Any discovered fans will be stopped and removed from memory.
        """
        self.stop()
        for fan in self._devices:
            fan.stop()
        self._devices = []


async def discover_any(timeout_seconds=5) -> bool:
    """Return True if any SenseME fans are found on the network.

    This function will always take timeout_seconds to complete.
    This method is a coroutine.
    """
    discovery = SensemeDiscovery(True, 1)
    discovery.start()
    await asyncio.sleep(timeout_seconds)
    count = len(discovery.devices)
    discovery.stop()
    _LOGGER.debug("Discovered %s fan%s", count, "" if count == 1 else "s")
    return count > 0


async def discover(value, timeout_seconds=5) -> SensemeFan:
    """Discover a fan with a fan name, room name or IP Address.

    None is returned if the fan was not found.
    This function will take up timeout_seconds to complete.
    This method is a coroutine.
    """
    start = time.time()
    discovery = SensemeDiscovery(True, 1)
    try:
        discovery.start()
        while True:
            await asyncio.sleep(0.1)
            devices = discovery.devices.copy()
            for device in devices:
                if device == value:
                    return device
            if time.time() - start > timeout_seconds:
                return None
    finally:
        discovery.stop()


aiosenseme/scripts/commandline.py
"""Script interface for aiosenseme library."""

import argparse
import asyncio
import logging
from typing import List

import aiosenseme
from aiosenseme import SensemeDiscovery, SensemeFan
from aiosenseme import __version__

ARGS = argparse.ArgumentParser(
    description="Discover and control Haiku/SenseME fans by Big Ass Fans."
)
ARGS.add_argument(
    "-V",
    "--version",
    action="store_true",
    dest="version",
    default=False,
    help="display version number",
)
ARGS.add_argument(
    "--debug",
    action="store_true",
    dest="debug",
    default=False,
    help="enable debug level logging",
)
ARGS.add_argument(
    "-d",
    "--discover",
    action="store_true",
    dest="discover",
    default=False,
    help="discover all fans on the network",
)
ARGS.add_argument(
    "-m",
    "--models",
    action="store_true",
    dest="models",
    default=False,
    help="list known fan models",
)
ARGS.add_argument(
    "-n",
    "--name",
    action="store",
    dest="name",
    default=None,
    help="fan name",
)
ARGS.add_argument(
    "-i",
    "--ip",
    action="store",
    dest="ip",
    default=None,
    help="fan ip address",
)
ARGS.add_argument(
    "-p",
    "--fan",
    action="store",
    dest="fan",
    default=None,
    choices=["on", "off"],
    help="fan power",
)
ARGS.add_argument(
    "-s",
    "--speed",
    action="store",
    dest="speed",
    default=None,
    type=int,
    choices=range(0, 8),
    help="fan speed",
)
ARGS.add_argument(
    "-l",
    "--light",
    action="store",
    dest="light",
    default=None,
    choices=["on", "off"],
    help="light power",
)
ARGS.add_argument(
    "-b",
    "--brightness",
    action="store",
    dest="brightness",
    default=None,
    type=int,
    choices=range(0, 17),
    help="light brightness",
)
ARGS.add_argument(
    "-w",
    "--whoosh",
    action="store",
    dest="whoosh",
    default=None,
    choices=["on", "off"],
    help="whoosh mode",
)

# array of discovered devices
_DEVICES = []


def print_fan(fan: SensemeFan):
    """Print information about fan."""
    msg = f"{fan.name}\n"
    if fan.room_status:
        msg += f"  Room Name: {fan.room_name}, Room Type: {fan.room_type}\n"
    if fan.has_light:
        msg += f"  Model: {fan.model} with light, "
    else:
        msg += f"  Model: {fan.model} without light, "
    msg += f"FW Version: {fan.fw_version}\n"
    msg += f"  IP Addr: {fan.ip}, MAC Addr: {fan.mac}\n"
    msg += f"  Token: {fan.network_token}"
    print(msg)


def print_fan_state(prefix: str, fan: SensemeFan):
    """Print information about fan current state."""
    msg = prefix
    if fan.fan_on:
        msg += f": Fan is on (speed: {fan.fan_speed})"
    else:
        msg += ": Fan is off"
    if fan.light_on:
        msg += f", Light is on (brightness: {fan.light_brightness})"
    else:
        msg += ", Light is off"
    if fan.fan_whoosh:
        msg += ", Whoosh: on"
    else:
        msg += ", Whoosh: off"
    print(msg)


async def discovered(fans: List[SensemeFan]):
    """Discovered fan callback function.

    Called when discovery has detected a SenseME fan.
    Each time a device is discovered all devices discovered are reported.
    """
    global _DEVICES
    for fan in fans:
        if fan not in _DEVICES:
            _DEVICES.append(fan)
            print_fan(fan)


async def process_args():
    """Process command line arguments."""
    args = ARGS.parse_args()
    if args.debug is True:
        logging.basicConfig(level=logging.DEBUG)
    if args.version is True:
        print("aiosenseme %s" % __version__)
        print("Copyright (C) 2020 by Mike Lawrence")
        print(
            "This is free software. You may redistribute copies of it under the terms"
        )
        print(
            "of the GNU General Public License <http://www.gnu.org/licenses/gpl.html>."
        )
        print("There is NO WARRANTY, to the extent permitted by law.")
    elif args.discover is True:
        try:
            discovery = SensemeDiscovery(True, 1)
            discovery.add_callback(discovered)
            discovery.start()
            await asyncio.sleep(4)
        finally:
            discovery.stop()
    elif args.models is True:
        msg = "Known fan models: "
        first = True
        for model in SensemeFan.models():
            if first:
                first = False
                msg += model
            else:
                msg += ", " + model
        print(msg)
        return
    else:
        if args.name is None:
            print("You must specify a fan name using -n or --name")
            return
        if args.ip is None:
            print("You must specify a fan ip address using -i or --ip")
            return
        fan = SensemeFan(args.name, "01:23:45:67:89:AB", args.ip, "FAN,HAIKU,SENSEME")
        fan.start()
        await asyncio.sleep(1.0)
        if fan is None:
            print(f"Fan/Room/IP address '{args.name}' not found")
            return
        print_fan(fan)
        print_fan_state("State", fan)
        try:
            changed = False
            if args.whoosh is not None:
                print(f"whoosh={args.whoosh}")
                if fan.fan_whoosh != (args.whoosh == "on"):
                    changed = True
                fan.fan_whoosh = args.whoosh == "on"
            if args.speed is not None:
                if args.fan is not None:
                    print(
                        "When specifying --fanspeed there is no " "reason to set --fan"
                    )
                if fan.fan_speed != args.speed:
                    changed = True
                fan.fan_speed = args.speed
            elif args.fan is not None:
                if fan.fan_on != (args.fan == "on"):
                    changed = True
                fan.fan_on = args.fan == "on"
            if args.brightness is not None:
                if args.light is not None:
                    print(
                        "When specifying --brightness there is no "
                        "reason to set --light"
                    )
                if fan.light_brightness != args.brightness:
                    changed = True
                fan.light_brightness = args.brightness
            elif args.light is not None:
                if fan.light_on != (args.light == "on"):
                    changed = True
                fan.light_on = args.light == "on"
            if changed:
                await asyncio.sleep(0.5)
                print_fan_state("New State", fan)

        finally:
            fan.stop()


def cli():
    """Command line interface for SensemeDiscovery."""
    task = asyncio.Task(process_args())
    loop = asyncio.get_event_loop()
    loop.run_until_complete(task)
